### PR TITLE
Master And Necromancer Zombie Improvements/Changes [READY] (I hope)

### PR DESCRIPTION
**Pull request by i2amroy**
1) Both the master and the necro now hang back tracking while there are nearby zombies that they could upgrade/revive and charge you when there isn't anything they could boost nearby (instead of the master charging you always, and the necro ignoring you completely).
2) The master zombie now uses the new "upgrades_into" system for it's upgrade choices, rather than just upgrading basic zombies.
3) The master zombie got it's upgrade range doubled and time cost slightly decreased, but it's cooldown time more than tripled.
4) Necromancer special has been reworked:

```
Old:
Necromancers could revive multiple zombies at once.
Each time a Necromancer revives a zombie it would lose a random amount of speed
Once a Necromancer's speed was halved it no longer revived zombies.
Special attack use: Every 4th turn
Range: 4

New:
Necromancers revive one zombie at a time
Each time a Necromancer revives a zombie it loses 20% of it's maximum speed
Each turn a Necromancer has a normalized chance of regaining lost speed up to its maximum. Speed is regained on average at 10% per 1.5 minutes
Special attack use: Every turn
Range: 10
```

The result is that a necromancer gets a fairly big ability to burst revivals at the start of the fight, but after you've been fighting a while (or if you can take out multiple zombies simultaneously) they run out of juice and need to recharge (and in addition the more "out of juice" they are the less they will move around due to their slower speed). I tried to balance it so that in general you are going to be better off trying to ignore the current nearby zombies to take down the necromancer first, but it's still possible for a person with an average weapon and a bit of melee skill to eventually bash their way through the wall of zombies being revived (though if you ever run into 2 necros at once you're going to be fighting for quite a long time before they both run out of juice, and with 3 you probably are going to either need to lure them apart or use some sort of explosive to take out all 3 at once).

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Master forcing zeds into next upgrade type permanently sounds like something that could easily wreck balance and also result in heavy metagame (leave master-contaminated areas until you can rush in).

Consider the new zombies like the predator. Predator is noticeably stronger than a hulk and also really hard to avoid. Its balancing relies on it not spawning early on.
If master can in just 40 turns turn a regular zed into a predator, it means you'd have to run into cities full speed, scout ahead to check for masters and immediately run away to despawn them if any are there.

No problem with the necromancer. To make it even better, you could give it a secondary ability to heal a nearby zed and restore it speed when it can't rez anything. Or maybe give this ability to master? Or maybe to some other new zed?

**Comment by i2amroy** [comment]
>Master

Yeah, that was something I was definitely kinda worried about, but I can't think of any good way to limit it. Maybe bump the cooldown up by a bunch more? The old master upgrade literally just duplicated the upgrade ability for normal zeds, so if possible I'd like to keep it tied into our new upgrade system for consistency.

> Necro

Definitely a plausible idea, I'll take a look into it. Probably fits the necro more than the master, especially considering zombies normally take a speed/hp penalty when they get revived anyways.

**Comment by Coolthulhu** [comment]
For now I'd rather keep masters inconsistent with the upgrade system and instead more consistent with what happens when the player isn't around.

To make it balanced, some sort of strength measure could be introduced. Say, regular zed is level 0, runner is lvl 1, hunter lvl 2 and predator lvl 3. Master could only bump it to the level equal to number of upgrade half-lifes that passes + 1. It could be measured by zeds' difficulty stat, but then those would have to be balanced properly because they're wildly inconsistent at the moment.

**Comment by i2amroy** [comment]
Expounding on my earlier idea, I think it might be pretty simple to just have it essentially give a random zombie nearby it the equivalent of one or two extra "days" of upgrading each time it used it's special (just backdate the last updated number of the zed to the day prior, then call the upgrade check function). That shouldn't be too powerful, while it could still allow it to hypothetically upgrade monsters. (And it would still respect the minimum upgrade dates). I'll have to double check, but I think the idea could work out to be pretty viable without making masters leave behind horrible trails of endgame enemies.

**Comment by Coolthulhu** [comment]
This sounds like something that doesn't have a large effect until it accumulates. Basically the opposite of how I'd like it to be - strong players who can rush in would be unaffected, weak players would benefit heavily from metagaming the master away.

**Comment by i2amroy** [comment]
You could hypothetically bump it up to a week or two of "upgrade time" it adds per use. For example if we set the upgrade time to 2 weeks, that would mean that 50% of the time a single use would result in a normal zombie instantly upgrading, while if an early game player runs into it it's not going to result in a sudden surge of predators while a later game player might. (And honestly with the current upgrade system a late game player isn't going to have any normal zeds around for a master to upgrade anyways).

It also has the advantage of not needing to add another field or token to every single possible upgradable monster in the game, and then hardcode limits based on them, and it's effect would automatically scale with difficulty as well.

**Comment by Coolthulhu** [comment]
> You could hypothetically bump it up to a week or two of "upgrade time" it adds per use.

You mean like limiting it to two weeks per zed or just having it do the "upgrade dance" slower and punishing the players who didn't scout ahead to despawn any potential masters?

If it reflected the upgrade by decrementing zed's upgrade counter, then it could work.

For example: Zed upgrades once per 2 weeks. 1 week old zed is upgraded by a master. It now needs 3 weeks to reach next upgrade stage and the master can't help it until it is at most 2 weeks from the next upgrade time.

But then you'd still have to solve the early predators. Possibly by bumping their upgrade time or by giving each zed type independent "shouldn't appear before upgrade time x" stat.

**Comment by i2amroy** [comment]
Not quite. This is basically the process I'm describing:
1) When zombie upgrades it checks the time since it's last update, figures out how many half lives have passed since that point, and then rolls an upgrade chance based on that number.
2) Zombie master uses his ability on a nearby zombie. The zombies current "last update roll" is the current day, since it was updated when it last loaded. The master set's that "last update roll"  number to a date two weeks prior to the current one. The upgrade function is then called on the zombie.
3) The zombie then checks it's "last update roll" number, and thinks that it's been two weeks since it last rolled to update. It then makes a roll against it's half life to see if it should upgrade, and should it pass it is automatically upgraded. This roll respects the "minimum upgrade time", meaning that if the zombie master were to try to upgrade something into a predator (which have a minimum upgrade time of day 14) prior to day 14 it would automatically fail (and it's relatively easy to just exclude any such cases from the master's targeting algorithm altogether.
4) Should the zombie upgrade (it's name will change) then the master special then prints a descriptive message about it upgrading one zombie into another.

The end result is that if you fight masters before they would normally start showing up, the master can't upgrade things into them. Meanwhile 14 days is a long enough time period to lead to pretty reliable upgrades on most zombies, normal zeds, with a half life of 14 days, will have a 50% chance of upgrading each time the master hits them with it's special.

**Comment by Coolthulhu** [comment]
That sounds mostly OK. One thing I don't like is how the metagame is still there (master will only start upgrading things when it spawns rather than having all nearby zeds spawn upgraded in his presence), but this version doesn't break the balance.

Eventually it would be nice for master to upgrade zeds offscreen and also have a meaningful ability to use on upgraded zeds (some sort of temporary buff), but for now this sounds OK.

**Comment by i2amroy** [comment]
>Eventually it would be nice for master to upgrade zeds offscreen and also have a meaningful ability to use on upgraded zeds (some sort of temporary buff), but for now this sounds OK.

Yeah, that's definitely on my list of things to look into once we actually have some form of "process mongroup while the player isn't here" type thing (at which point we'll probably drop the power of the upgrade ability fairly drastically to compensate for it working all the time).

As for a temporary buff, the eventual plan is to have the master passively increase the pathfinding intelligence of nearby zeds, sort of similar to the way the brain blob does with blobs ATM, but on a bit higher level. For example a group of shockers near the master might hold back lobbing shockbolts while a few of the more tanky zeds act as a wall to stop you from reaching them; the master might chase you down with one group while sending another group of zombies in an attempted flanking maneuver; or the master might grab a bunch of spitters and have them target the area around you to pin you down, while the zombies form a big circle around the puddle and watch you slowly melt; that sort of stuff. It's one of the rare zeds that is pretty much outright stated to have some intelligence as well as being a "master" of others, and we intend to milk that idea for all that it's worth :P. (Balancing will obviously be done when we get around to actually adding it).

**Comment by i2amroy** [comment]
The healing special idea didn't work out so well, necros ended up blowing most of their juice on healing living monsters and then became unable to revive them after they were dead. I'm thinking what I'll do instead is make a follow-up PR that pulls masters and Necros into a "support-type zombie" upgrade chain, along with a few other zombie ideas as well (healer zed? Short term speed booster zed? I'm kicking around a few different ideas). There would also probably be some weaker forms of the different support zeds that would spawn mainly at the beginning, then upgrade into their later types.

Also took the opportunity to quick fix the difficulty scaling with monster upgrading times; this is ready for merging since I'll do the actual support zed stuff in a later PR where we can debate their finer points without bogging this particular rebalance down.

**Comment by Coolthulhu** [comment]
You could have them heal for free, but in some low amounts.

As for the monster upgrade times: it would be cool if someone reversed the difficulty setting so that low means low difficulty and 0 means no upgrading at all.

**Comment by i2amroy** [comment]
Ah, should be fairly easy for me to fix up the difficulty setting (And fix the divide by zero I accidentally just put into the PR).

And I'll almost certainly have something to do with the necros in the support PR, but at this point the ideas and implementations just aren't quite clicking in my head like they should. For now I think I'm probably just going to leave them like this (since what is already there is basically guaranteed to still be there in the support PR, and is already an improvement over the current necromancers) and let the ideas bounce around in my head for another couple days. Be assured that I *will* be adding something to fill in the semi-missing spot necro's currently have, but I figure it's probably ok to leave it unfilled (since it's not filled now anyways) for a few more days while I hash out exactly the direction that I want to aim them in. At absolute worst (and assuming I don't commit Seppuku over my upcoming finals) I should have something up in a little over a week that should fill in any missing holes that this one leaves unfilled.

**Comment by i2amroy** [comment]
There, monster difficulty now works in a sensible fashion where higher values make monsters take longer to upgrade and lower ones make them take longer; with zero stopping all upgrading where possible.

**Comment by Coolthulhu** [comment]
The difficulty inversion will require more than just a division. We need to handle the old saves somehow.
The best way would probably be renaming the world option and setting it up using the old one. Also the description has to change.

The master seems to have a hardcoded upgrade time of 2 weeks. That should probably be scaled with difficulty - both to mean something on lower difficulties and not to result in predohulk spam on higher ones.

I'll mergetest later if those are resolved by the time I push my PR.

And a suggestion:
You could give master the swarming tag from bees/wasps - it would make it seek out other zeds when it has nothing to do.

**Comment by i2amroy** [comment]
Not gonna happen tonight, it's 3:30 am here already and I don't have near the time required to figure out why a simple name and description change of the option isn't properly adding it to older world saves like the code seems to imply it should. Might be able to get to it tomorrow or Thursday, we'll see, but it's definitely not happening tonight.

**Comment by kevingranade** [comment]
Perhaps we want some kind of down side to the upgrade or other rationale
for why the master isn't simply surrounded by more advanced zombies all the
time.  If we had the master upgrade nearby zombies in the background, but
made the ability respect some exclusion date (which I'm against by the way,
there's no need for arbitrary limits on when monster types appear),
wouldn't the nearby zombies just stay at their limits all the time?
As for implementation, with the proposed system, if you set the last
upgrade check time to the past, you could then set the check time to the
future if it was successful, this would also tend to spread out the
upgrades to all nearby zombies and eventually cap instead of everything
eventually progressing to being a hulk or similar.

I'm looking on some clarification of the 'despawning masters' thing though,
do you mean the player always clears out the other zombies before engaging
the master?  If so I fail to see the problem, especially since it's
something you'd do with any dangerous monster, and it has multiple limiters
on it we want to add anyway.
1. With hordes 2.0, just-off-the-screen monsters will be pulled in my any
particularly noisy altercations, making pulling dangerous monsters more
difficult.
2. I keep planning on adding a small noise to zombies when they are moving
toward something, which will tend to pull the rest of the zombies with them.

**Comment by Coolthulhu** [comment]
> I'm looking on some clarification of the 'despawning masters' thing though,
do you mean the player always clears out the other zombies before engaging
the master?

No, I meant something like player approaching the master in such a way that either only masters or only zeds are spawned, so that the master can't upgrade the zeds even though they are nearby.

> there's no need for arbitrary limits on when monster types appear

I can't think of a better idea than limiting appearance time. Having them all spawn without limitations would lead to very hard early game and very easy late game.
Unless you mean soft limits (+/- x days ), which would be workable.

> wouldn't the nearby zombies just stay at their limits all the time?

Most zeds don't transform immediately when it is possible, but some time after. Masters could shorten this extra time needed and always be surrounded by a bunch of elite guards, a level above regular zombies.

File: src/monattack.cpp [hunk 1]
```diff
@@ -3499,37 +3521,57 @@ void mattack::generator(monster *z, int index)
...
+            add_msg(m_warning, _("A black mist floats from the %1$s around the %2$s."),
+                     z->name().c_str(), old_name.c_str());
+        } else {
+            add_msg(m_warning, _("A black mist floats from the %s."), z->name().c_str());
+        }
+    }
+    if (target->name() != old_name) {
+        if( could_see && can_see ) {
+            //~ %1$s is the pre-upgrade monster, %2$s is the post-upgrade monster.
+            add_msg(m_warning, _("The %1$s becomes a %2$s!"), target->name().c_str());
```
Comment (BevapDin) [inline src/monattack.cpp]: I see a %2$s, but there is only one further argument, this will crash.

File: src/monster.cpp [hunk 2]
```diff
@@ -1652,6 +1658,16 @@ void monster::make_ally(monster *z) {
     faction = z->faction;
 }
 
+int monster::get_last_load()
```
Comment (BevapDin) [inline src/monster.cpp]: Should be const.

File: src/monattack.cpp (hunk 1 above)
Comment (Zireael07) [inline src/monattack.cpp]: I think you're mising old_name.c_str() in there...

**Comment by kevingranade** [comment]
> No, I meant something like player approaching the master in such a way
that either only masters or only zeds are spawned, so that the master can't
upgrade the zeds even though they are nearby.

Unless the player can engage the master or zombies while said targets are
in the outermost ring of map chunks (~12 tiles from the border of the
bubble), there's no way to do that, and if that is the case they either
deserve it because they earned that ability, or there's an ability that
needs nerfing.  Also keep in mind what I said about loud noises reaching
outside the reality bubble, most attacks with that kind of range should be
very noisy.

>> there's no need for arbitrary limits on when monster types appear
>
> I can't think of a better idea than limiting appearance time. Having them
all spawn without limitations would lead to very hard early game and very
easy late game.
> Unless you mean soft limits (+/- x days ), which would be workable.

I specifically mean having monsters decay to new types in a way very
similar to radioactive decay, and spawning only some base types at the
start of the game.  What I don't want is limits like, 'monster x can't
upgrade to monster y until two weeks after game start'.

>> wouldn't the nearby zombies just stay at their limits all the time?
>
> Most zeds don't transform immediately when it is possible, but some time
after. Masters could shorten this extra time needed and always be
surrounded by a bunch of elite guards, a level above regular zombies.

It seems like if you have masters upgrading nearby zombies both during
combat and out of combat, you're either going to have in-combat upgrading
not happening much, or the effects are going to be stacking too much.
Given a choice between in-combat upgrades and off screen upgrades, I much
prefer the former, because players can interact with it and see what's
happening.
As for your point about keeping interactions with the master consistent, as
long as the ability rolls forward the 'last upgrade check' into the future,
the master will never be able to advance monsters past a certain offset in
the future, no matter how often or long the player interacts with them.

**Comment by vache** [comment]
On a semi-related note, there could be some sort of flavor based delay to the master's upgrade ability.  Like, it triggers it on a nearby zombie, the zombie gets stunned for 500 turns, the message log says something about the zombie getting all lumpy or something, then boom, new upgraded zombie.

**Comment by i2amroy** [comment]
Seems as if the crashes I was getting were remnants of and old build; upon this latest rebuild they have vanished completely and this seems to be working fine.

As for the option in old worlds, it's using the "defaults fallback" that was already implemented. If a world doesn't have a setting (which is the case for any old worlds) it will use whatever the current setting is in the options menu instead. Doesn't 100% totally fix the problem, but anything more would require me to crack open the entire saving/loading options process as best as I can tell, which is beyond the scope of this PR (and if it's really that important I'd just revert the option changes out of this PR).

File: src/monattack.cpp [hunk 3]
```diff
@@ -3462,38 +3484,59 @@ void mattack::generator(monster *z, int index)
...
-    const auto monsters = MonsterGroupManager::GetMonstersFromGroup("GROUP_ZOMBIE_UPGRADE");
-    const std::string newtype = monsters[rng(0, monsters.size() - 1)];
-
+    std::string old_name = target->name();
     const auto could_see = g->u.sees( *target );
-    target->poly(GetMType(newtype));
-    const auto can_now_see = g->u.sees( *target );
+    // Currently gives zombies the equivalent of a week of upgrade time
+    // Difficulty scaling happens in update_check()
+    target->set_last_load(target->get_last_load() - 7);
```
Comment (Coolthulhu) [inline src/monattack.cpp]: If I understand it right, master, if given time, will upgrade all nearby zeds to the best form allowed by current time.
Is this how is it supposed to work?

File: src/monattack.cpp [hunk 4]
```diff
@@ -3462,38 +3484,59 @@ void mattack::generator(monster *z, int index)
...
+    int day = calendar::turn / DAYS(1);
     for (size_t i = 0; i < g->num_zombies(); i++) {
         monster &zed = g->zombie(i);
-        if( zed.type->id == "mon_zombie" &&
-            rl_dist( z->pos(), zed.pos() ) <= 5 &&
-            z->sees( zed ) &&
-            z->attitude_to( zed ) != Creature::Attitude::A_HOSTILE ) {
-            targets.push_back(i);
+        // Check this first because it is a cheap check
+        if( zed.type->upgrade_min < day &&
```
Comment (Coolthulhu) [inline src/monattack.cpp]: `upgrade_min` is not scaled to difficulty, neither is `day`. `monster::update_check` however does scale `upgrade_min`.

Comment (i2amroy) [inline src/monattack.cpp]: Good catch, will have a fix up in just a second.

File: src/monattack.cpp (hunk 3 above)
Comment (i2amroy) [inline src/monattack.cpp]: At the moment yes, but it may take several uses on each individual zombie to actually get them to that stage.

**Comment by Coolthulhu** [comment]
Minor issues:
* Masters and necros who lose their anger don't lose their plans, meaning they will move in the old direction (charge the nearby player/animal) unless something stops them
* Necromancer can rez unseen corpses and can rez them through transparent obstacles as well
* Necros won't fight back (not that they could do much) or run if there's even a single zed around. This makes them pretty much always docile.
* The code for corpse finding for necros will always check all tiles in radius, even when it only wants to check if it should get angry (needs only 1 corpse for that)
* In case of a failed revive (for example, burned corpse), there will be no message that the rezing was attempted

The only major issue I see is the whole metagame thing with masters only working when spawned.

**Pull request merged**